File: app/src/main/java/com/iamdsr/travel/ViewModels/PlanTripFragmentViewModel.java
package com.iamdsr.travel.ViewModels;

import android.util.Log;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.iamdsr.travel.Interfaces.FirestoreTaskCompleteInterface;
import com.iamdsr.travel.Models.TripModel;
import com.iamdsr.travel.Repos.FirebaseRepositories;

import java.util.List;

public class PlanTripFragmentViewModel extends ViewModel implements FirestoreTaskCompleteInterface {

    private MutableLiveData<List<TripModel>> plannedTripListMutableLiveData;
    private FirebaseRepositories firebaseRepositories = new FirebaseRepositories(this);

    public PlanTripFragmentViewModel(){
        if (plannedTripListMutableLiveData == null) {
            plannedTripListMutableLiveData = new MutableLiveData<>();
            firebaseRepositories.getPlannedTripsFromDB();
        }
    }
    public LiveData<List<TripModel>> getPlannedTripListMutableLiveData() {
        //Log.d("TAG", "getPlannedTripListMutableLiveData: "+plannedTripListMutableLiveData);
        return plannedTripListMutableLiveData;
    }

    public void updateTrips(List<TripModel> list){
        firebaseRepositories.getUpdatedPlannedTripsFromDB(list);
    }
    @Override
    public void onNewPlansAdded(List<TripModel> tripModelList) {
        //Log.d("TAG", "onNewPlansAdded: "+tripModelList.size());
        plannedTripListMutableLiveData.setValue(tripModelList);
        //Log.d("TAG", "onNewPlansAdded: "+plannedTripListMutableLiveData);
    }

    @Override
    public void onPlanUpdated(List<TripModel> tripModelList) {
        plannedTripListMutableLiveData.setValue(tripModelList);
    }

    @Override
    public void onException(Exception exception) {

    }
}

File: app/src/main/java/com/iamdsr/travel/viewModels/PlanTripFragmentViewModel.kt
package com.iamdsr.travel.viewModels

import android.content.ContentValues.TAG
import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.google.firebase.firestore.EventListener
import com.google.firebase.firestore.Query
import com.google.firebase.firestore.QuerySnapshot
import com.iamdsr.travel.models.TripModel
import com.iamdsr.travel.repositories.FirestoreRepository

class PlanTripFragmentViewModel : ViewModel() {

    var firebaseRepository = FirestoreRepository()
    var savedTrips : MutableLiveData<List<TripModel>> = MutableLiveData()

    // save address to firebase
    fun saveNewTripToFirebase(tripModel: TripModel){
        firebaseRepository.addNewPlannedTripToDB(tripModel).addOnFailureListener {
            Log.e(TAG,"Failed to save Trip!")
        }
    }

    // Update trip
    fun updateTripToFirebase(tripMap: MutableMap<String, Any>, tripID: String){
        firebaseRepository.updateExistingTripToDB(tripMap, tripID).addOnFailureListener {
            Log.e(TAG,"Failed to update Trip!")
        }
    }

    // get realtime updates from firebase regarding saved addresses
    fun getSavedTrips(): LiveData<List<TripModel>> {
        firebaseRepository.getSavedTrips().orderBy("date_created", Query.Direction.DESCENDING).addSnapshotListener(EventListener<QuerySnapshot> { value, e ->
            if (e != null) {
                //Log.w(TAG, "Listen failed.", e)
                //savedTrips.value = null
                return@EventListener
            }

            val savedTripsList : MutableList<TripModel> = mutableListOf()
            for (doc in value!!) {
                val plannedTripItem = doc.toObject(TripModel::class.java)
                savedTripsList.add(plannedTripItem)
            }
            savedTrips.value = savedTripsList
        })
        return savedTrips
    }
}